### ISSUE TITLE
Custom Callbacks for ClearML Failing During Multi-GPU Training

### ISSUE THREAD

**Issue by 421psh** [ultralytics/ultralytics#18070]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

When executing training tasks on multiple GPUs, the custom callbacks integrated with ClearML are not functioning as expected. This issue seems to arise specifically during multi-GPU operations, while single GPU training does not exhibit the same problem. When setting `device=0`, the callbacks function correctly. However, if you set `device=0,1,2,...`, the callbacks do not work. Perhaps such behavior of callbacks without ClearML is the same. The training process is executed by the ClearML agent on a computing server during the ClearML pipeline.

A similar issue has been discussed here [#6913](https://github.com/ultralytics/ultralytics/issues/6913). Although adding a RANK check to the code seemed to help in the past, it no longer resolves the issue.

### Environment

Ultralytics 8.3.0 🚀 Python-3.12.3 torch-2.4.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3070 Ti, 7869MiB)
Setup complete ✅ (16 CPUs, 31.3 GB RAM, 51.1/93.2 GB disk)

OS                  Linux-6.8.0-49-generic-x86_64-with-glibc2.39
Environment         Linux
Python              3.12.3
Install             pip
RAM                 31.27 GB
CPU                 AMD Ryzen 7 5800X 8-Core Processor
CUDA                12.1

numpy               ✅ 1.26.4<2.0.0,>=1.23.0
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.4.1>=1.8.0
torchvision         ✅ 0.19.1>=0.9.0
tqdm                ✅ 4.66.5>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.6>=2.0.0
torch               ✅ 2.4.1!=2.4.0,>=1.8.0; sys_platform == "win32"

### Minimal Reproducible Example

```
from clearml import Task
from ultralytics import YOLO
from ultralytics.utils import RANK


def checkpoint_save_callback(trainer):
    if RANK in {-1, 0}:
        if task := Task.current_task():
            task.update_output_model(
                model_path=str(trainer.last),
                model_name="last",
                auto_delete_file=False,
                iteration=trainer.epoch,
            )
            if trainer.save_period > 0 and trainer.epoch % trainer.save_period == 0:
                model_name = f"epoch{trainer.epoch}"
                task.update_output_model(
                    model_path=str(trainer.last.with_stem(model_name)),
                    model_name=model_name,
                    auto_delete_file=False,
                    iteration=trainer.epoch,
                )


def train(model_path: str, config: dict) -> None:
    model = YOLO(model_path)
    if config.get("save_period", -1) > 0:
        model.add_callback("on_model_save", checkpoint_save_callback)
    model.train(**config)
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @421psh, thank you for your interest in Ultralytics 🚀! We recommend visiting the [Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, thank you for providing extensive details about the issue and your environment 🙌. To proceed, we kindly request a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) (MRE) to better understand and debug the problem, if the current code snippet does not fully encapsulate the issue. Ensuring that the issue is replicated within a simplified setup will help us investigate more effectively.

If this is related to custom training ❓, please ensure you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Providing additional context such as logs or any further observations is also highly encouraged.

Join the Ultralytics community where it suits you best! For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Before proceeding, please verify your issue persists with the latest `ultralytics` package and dependencies. You can upgrade in a [**Python>=3.8**](https://www.python.org/) environment with:

```bash
pip install -U ultralytics
```

This ensures that the issue is not already resolved in the latest release.

## Environments

Ensure that you are using a supported environment for multi-GPU training. YOLO performs best in verified environments, for example:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response to help expedite your issue. Rest assured, an Ultralytics engineer will review your issue and provide further assistance soon 😊.

**Comment by Y-T-G** [comment]
Custom callbacks don't work for multi-GPU training. They run in a separate DDP subprocess and the callbacks can't be passed to that process.

**Comment by Laughing-q** [comment]
@421psh Hey currently based on the mechanism we have for DDP training we don't support custom callback yet with it.
For now if you want to use custom callback in your DDP training you could directly launch the training script by:
```bash
python -m torch.distributed.launch --nproc_per_node=num_gpu train.py
```

**Comment by 421psh** [comment]
@Y-T-G @Laughing-q, thank you for the answers. In my case, running through `torch.distributed.launch` is not a suitable option, as train is one of the components of the **ClearML** `PipelineDecorator`. Am I correct in assuming that we shouldn't expect this issue to be fixed in the **Ultralytics** repository until there's a solution from **torch** itself? So should I close this issue?

**Comment by Laughing-q** [comment]
@421psh Well I think it's a more Ultralytics issue than torch, it's because we are handling the DDP launching internally by generating a DDP command and launch it internally:
https://github.com/ultralytics/ultralytics/blob/e0f8eda366c7cf08a1c311a17eaaaf87914f8012/ultralytics/utils/dist.py#L65
so users don't need to explicitly use `torch.distributed.launch` in their command.
However since we launch the DDP command internally so it's hard to capture customized callbacks from users. One quick way to combine customized callbacks is to launch the script with `torch.distributed.launch` explicitly as I recommended before.

> In my case, running through torch.distributed.launch is not a suitable option, as train is one of the components of the ClearML PipelineDecorator.

Do you have a training script as entrance point to launch the training? if so you could just need to explicitly add `torch.distributed.launch` and it'll work to you.

**Comment by Y-T-G** [comment]
You can also try this branch. It uses dill to save the callbacks and loads it in the subprocess.

https://github.com/ultralytics/ultralytics/pull/18081

You can install it using:
```bash
git clone https://github.com/Y-T-G/ultralytics/tree/cb_ddp
cd ultralytics
git checkout cb_ddp
pip install .
```

Then run training as usual.

**Comment by 421psh** [comment]
@Y-T-G Great, I'll wait for your changes to be merged into the main repository. I am looking forward to it.

**Comment by Y-T-G** [comment]
You could test it and see if it works. I'm not sure if it works for complex callbacks, especially if it involves using sessions or other states that are in the current process. They may not be transferrable to a different process.
